TINKERPOP-1599 implement real gremlin-python driver

davebshow opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-1599

This PR adds a better driver implementation for gremlin-python:

- Uses a multi-threaded solution to provide asynchronous I/O
- Decouples the underlying websocket client implementation from driver code, making it easy to plug in a different client
- Makes it easy to plug in different protocols, like the Gremlin Server HTTP protocol.
- Adds simple connection pooling used in concurrent requests, which increases driver performance with slow traversals
- Improves driver tests by adding `pytest` fixtures and removing unneeded `unittest` code.

This driver still isn't full featured compared to the Java driver, for example, it doesn't implement `Cluster` to use multiple hosts. But, it is considerably better than the old implementation, and as this was becoming a big PR, I figure we can add more features with subsequent PRs. 

Also note, I know we are in code freeze right now, so I don't expect this to be reviewed/merged for 3.2.4, but I'm busy and I wanted to get this done so I can move on.

spmallette commented:
@davebshow now that code freeze is lifted i think you should probably rebase this PR. Also, two documentation related items:

1. This works needs a changelog entry - it was probably better you didn't add that yet since you would have had to move it anyway to 3.2.5.
2. This change seems sufficiently large and important that it should have some documentation on usage in the reference docs and should likely have an entry in the upgrade docs as well so that users know that there as an actual driver that can be used for python now.

davebshow commented:
Yeah I figured we would need to add docs. Where should I add to the reference docs? Maybe in [Gremlin Applications](https://github.com/apache/tinkerpop/blob/master/docs/src/reference/gremlin-applications.asciidoc) after the [Connecting via Java](https://github.com/apache/tinkerpop/blob/master/docs/src/reference/gremlin-applications.asciidoc#connecting-via-java) section?

spmallette commented:
I think that make sense - just add a "Connecting via Python" section perhaps.

davebshow commented:
Ok, I have now rebased ad added docs to both reference and upgrading. I also reviewed the old implementation and made a couple small fixes for consistency. I also added standard op processor message serialization. I think this one is getting close to being ready to go.

spmallette commented on CHANGELOG.asciidoc:
```diff
@@ -29,6 +29,11 @@ TinkerPop 3.2.5 (Release Date: NOT OFFICIALLY RELEASED YET)
+
+Improvements
+^^^^^^^^^^^^
+TINKERPOP-1599 implement real gremlin-python driver
```
you don't need to add those manually - they get added on release (we generate a report out of jira for it).

davebshow commented:
Well, I've made my last review of this code and pushed a bit of cleanup and a small fix. IMHO, this is ready to be merged. I'll wait to see if there are any more comments or feedback before upvoting.

spmallette commented:
All tests pass with `docker/build.sh -t -n -i`

VOTE +1

davebshow commented:
I think this is ready. Anyone want to be a third reviewer for this?

VOTE +1

okram commented:
VOTE +1.

davebshow commented:
Merged.
